core/src/validator.rs

```diff
@@ -1639,15 +1663,30 @@ impl<'a> ProcessBlockStore<'a> {
                 )
             });
 
+            if let Some(hard_fork_restart_slot) = maybe_cluster_restart_with_hard_fork(
+                self.config,
+                self.bank_forks.read().unwrap().root_bank().slot(),
+            ) {
+                // reconciliation attempt 2 of 2 with hard fork
+                // this should be #2 because hard fork root > tower root in almost all cases
+                reconcile_blockstore_roots_with_external_source(
+                    ExternalRootSource::HardFork(hard_fork_restart_slot),
+                    self.blockstore,
+                    &mut self.original_blockstore_root,
+                )
+                .unwrap_or_else(|err| {
+                    error!("Failed to reconcile blockstore with hard fork: {:?}", err);
+                    abort()
+                });
+            }
+
             *self.start_progress.write().unwrap() = previous_start_process;
         }
     }
-}
 
-impl<'a> From<ProcessBlockStore<'a>> for Tower {
```

Comment:
@mvines I think this isn't correct use of `From`, which is introduced at #23852 (https://github.com/solana-labs/solana/pull/23852/files#r902237878).

This is converting quite different things. ideally, `From` should be a cast for conceptually similar things. Also, this made code reading a bit hard because i can't easily spot when `ProcessBlockStore::process` is exactly executed now. (i.e. searching the correct `.into()` is hard)

it seems this trait workaround is only needed for this test [test_tvu_exit](https://github.com/solana-labs/solana/pull/19912/files#r902242642) and i think alternative approach should arguably be tolerable (still not ideal, i skipped proper introducing enum or trait just for this purpose...): https://github.com/solana-labs/solana/pull/19912/files#r902243960

btw, i didn't know there is an identity/no-op blanket `From`/`Into` impl provided here
https://doc.rust-lang.org/src/core/convert/mod.rs.html#559:

```rust
// From (and thus Into) is reflexive
#[stable(feature = "rust1", since = "1.0.0")]
#[rustc_const_unstable(feature = "const_convert", issue = "88674")]
impl<T> const From<T> for T {
    /// Returns the argument unchanged.
    fn from(t: T) -> T {
        t
    }
}
```


granted, this refactor isn't relevant to this pr but i couldn't resist the urge, when i was code-reading to guarantee the safety of the scary blockstore mutation with hard fork params.

happy to separate this change into another preparatory pr, if you desire. :)

Reply:
Sure, this looks fine to me